src/lib/storage/fsst_segment.cpp

```diff
@@ -0,0 +1,167 @@
+
+  // Third party library call.
+  size_t output_size_after_decompression =
+      fsst_decompress(&_decoder, compressed_length, compressed_pointer, output_size, output_buffer.data());
+
+  // "shrink_to_fit" for output buffer.
+  output_buffer.resize(output_size_after_decompression);
+
+  pmr_string output{output_buffer.begin(), output_buffer.end()};
+  return {output};
+}
+
+template <typename T>
+ChunkOffset FSSTSegment<T>::size() const {
+  // Compressed offsets have size of values.size() + 1 (see fsst_encoder.hpp).
+  return static_cast<ChunkOffset>(_compressed_offsets->size() - 1);
+}
+
+template <typename T>
+std::shared_ptr<AbstractSegment> FSSTSegment<T>::copy_using_allocator(const PolymorphicAllocator<size_t>& alloc) const {
+  auto new_null_values =
+      _null_values ? std::optional<pmr_vector<bool>>{pmr_vector<bool>{*_null_values, alloc}} : std::nullopt;
+  auto new_compressed_values = pmr_vector<unsigned char>{_compressed_values, alloc};
+  auto new_compressed_offsets = _compressed_offsets->copy_using_allocator(alloc);
+  auto new_reference_offsets = pmr_vector<uint64_t>{_reference_offsets, alloc};
+  fsst_decoder_t new_decoder = _decoder;
+
+  auto copy = std::make_shared<FSSTSegment>(new_compressed_values, new_compressed_offsets, new_reference_offsets,
+                                            new_null_values, _number_elements_per_reference_bucket, new_decoder);
+  copy->access_counter = access_counter;
+
+  return copy;
+}
+
+template <typename T>
+size_t FSSTSegment<T>::memory_usage(const MemoryUsageCalculationMode) const {
+  auto compressed_values_size = _compressed_values.size() * sizeof(unsigned char);
+  auto reference_offsets_size = _reference_offsets.size() * sizeof(uint64_t);
+  auto compressed_offsets_size = _compressed_offsets->data_size();
+  auto null_value_size = (_null_values.has_value() ? _null_values->capacity() / CHAR_BIT : size_t{0});
+
+  return sizeof(*this) + compressed_values_size + compressed_offsets_size + reference_offsets_size + null_value_size;
+}
+
+template <typename T>
+EncodingType FSSTSegment<T>::encoding_type() const {
+  return EncodingType::FSST;
+}
+
+template <typename T>
+std::optional<CompressedVectorType> FSSTSegment<T>::compressed_vector_type() const {
+  return _compressed_offsets->type();
+}
+
+template <typename T>
+const fsst_decoder_t& FSSTSegment<T>::decoder() const {
+  return _decoder;
+}
+
+template <typename T>
+const pmr_vector<unsigned char>& FSSTSegment<T>::compressed_values() const {
+  return _compressed_values;
+}
+
+template <typename T>
+const std::unique_ptr<const BaseCompressedVector>& FSSTSegment<T>::compressed_offsets() const {
+  return _compressed_offsets;
+}
+
+template <typename T>
+const std::optional<pmr_vector<bool>>& FSSTSegment<T>::null_values() const {
+  return _null_values;
+}
+
+template <typename T>
+uint64_t FSSTSegment<T>::number_elements_per_reference_bucket() const {
+  return _number_elements_per_reference_bucket;
```

Comment:
Too many getters for my taste. Is there a reason that `_number_elements_per_reference_bucket` is not const?

Reply:
We need all getters for retrieving and saving all member variables in the `BinaryWriter`.